### PR TITLE
Alias support for `tsh`

### LINKED ISSUE #12341: Implement tsh command aliases
Implement tsh aliases proposed in [RFD 61](https://github.com/gravitational/teleport/blob/master/rfd/0061-tsh-aliases.md).

### PR DESCRIPTION
**Pull request by Tener**
This is a PR implementing RFD#61: [tsh command aliases](https://github.com/gravitational/teleport/blob/master/rfd/0061-tsh-aliases.md).

Notes:
- aliases are read from global and user configs.
- if possible, we call `Run()` instead of spawning fresh `tsh`. This reduces noise and leads to more consistent UX.
- alias loops are detected; this requires updated Kingpin (https://github.com/gravitational/kingpin/pull/8)
- if a parameter is referenced in an alias (e.g. `$0`, `$2`, ...) it is replaced within the invocation. Otherwise, it is added at the end of the command.
- We don't support `$@`. Argument handling as above supersedes it.
- if a referenced parameter is not found (too few arguments), an error is raised.
- flags before alias are moved; `tsh --foo --bar alias --bzz --zzz` becomes `tsh alias --foo --bar --bzz --zzz`.

For better alias UX a handful of other changes were made:
- added support for `TELEPORT_DEBUG`, changed how logging is initialized. debugging aliases is possible via `TELEPORT_DEBUG=1` env variable, --debug is ineffective.
- we prevent Kingpin from terminating the app, we are better at handling parsing errors ourselves.
- we re-enable interspersed flags/args, this works better with mixing the flag order.
- We enable `.AllRepeatable()` flag in Kingpin (https://github.com/gravitational/kingpin/pull/7). This makes it much easier to compose aliases.

Fixes #12341.

### CODE REVIEW

File: tool/tsh/aliases.go
Comment (smallinsky) [inline tool/tsh/aliases.go:32]: Can you provide some `tsh` tests for the full flow of  `tryRunAlias` function ?

File: lib/utils/cli.go
Comment (smallinsky) [inline lib/utils/cli.go:288]: I don't see any  the `ox.Exit` call added so in case of invalid flags or arguments like `tsh foo` the tsh will return `0` as exit status code ?

Comment (Tener) [reply lib/utils/cli.go:288]: No, the error code will be non-zero still. The thing is, without `app.Terminate(nil)` Kingpin will, in some cases, terminate the entire app instead of simply returning an error.

Comment (Tener) [reply lib/utils/cli.go]: Fixed in: dd26cf68919ebd92df3833a292958e597f79a85e

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go]: I've added test coverage. `tryRunAlias` no longer exists because of refactoring, but the equivalent functionality is now tested.

Comment (Tener) [reply tool/tsh/aliases.go]: I've added the struct to encapsulate the execution environment. One function initializes it for actual usage, the other does so for tests.

Comment (Tener) [reply tool/tsh/aliases.go]: Done.

File: tool/tsh/aliases.go
Comment (smallinsky) [inline tool/tsh/aliases.go:121]: nit: rm ? Personal I don't see any value in one word comments

File: tool/tsh/aliases.go
Comment (smallinsky) [inline tool/tsh/aliases.go:113]: nit: Can we reverse the returned args to "string"," bool" to be more complied with common GO convention ?

File: tool/tsh/tsh.go
Comment (smallinsky) [inline tool/tsh/tsh.go:483]: It is possible that the Terminate function is called severals time and first exitCode will be overwritten ?

File: tool/tsh/tsh.go
Comment (smallinsky) [inline tool/tsh/tsh.go:819]: nit: extract to a sperate function

**Review by smallinsky — COMMENTED** [review]
I have some comments but the code looks reasonable. We should take extra effort to ensure that we don't break any current `tsh` behavior. 

@r0mant Could you also take a look ?

File: tool/tsh/tsh.go
Comment (Tener) [reply tool/tsh/tsh.go:487]: In theory, yes. Not very likely, but I can ensure the non-zero exit code is sticky.

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go]: Done.

File: tool/tsh/tsh_test.go
Comment (Tener) [reply tool/tsh/tsh_test.go:151]: Added a similar test.

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:830]: Nit: I would combine this into alias runner's `Run()` method or something.

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:838]:
```suggestion
			log.Debugf("Failing due to recursive alias %q. Aliases seen: %v", aliasCommand, ar.getSeenAliases())
```

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:839]:
```suggestion
			return trace.BadParameter("recursive alias %q; correct alias definition and try again", aliasCommand)
```

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:482]: Is this needed to ensure that app.Parse() doesn't exit prematurely so we can return the recursive alias error? If so, I would move this closer to app.Parse() call and expanded the explanation, right now it's a bit confusing.

**Review by r0mant — APPROVED** [review]
Left a few questions/suggestions but approving so you're not blocked on me.

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go:41]: Done.

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go]: Added.

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go]: Moved.

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go]: I ended up rewriting this function, but it feels more solid now. Crucially `ReplaceAllStringFunc()` is now used for one-shot replacement instead of a loop, so you can now safely use "$123" as an argument to the alias; previously, you could not.

The new code also correctly handles variables above `$9`, such as `$10` or `$123`. Previous code would mishandle those in confusing ways (the error would have been unhelpful).

We note which variables are referenced and exist and which are referenced and do not exist (i.e., too few args).

File: tool/tsh/tsh.go
Comment (Tener) [reply tool/tsh/tsh.go]: Good point. Moved closer to where it is needed for easier comprehension.

File: tool/tsh/tsh.go
Comment (Tener) [reply tool/tsh/tsh.go]: Done.

File: tool/tsh/aliases.go
Comment (Tener) [reply tool/tsh/aliases.go:31]: Done.

**Review by smallinsky — APPROVED** [review]

File: tool/tsh/tsh.go
Comment (greedy52) [reply tool/tsh/tsh.go:481]: just FYI this broke `tsh aws` =p https://github.com/gravitational/teleport/issues/15965

**Pull request merged**
